File: One-to-One-Video/Meta-macOS-Tutorial-Objective-C-1to1/Meta-Mac-Tutorial-Objective-C/DeviceSelectionViewController.m
//
//  DeviceSelectionViewController.m
//  Meta Mac Tutorial Objective-C
//
//  Created by 3 on 2020/12/17.
//  Portions Copyright (c) 2020 meta-rti. All rights reserved.
//

#import "DeviceSelectionViewController.h"

@interface DeviceSelectionViewController ()

@property (weak) IBOutlet NSPopUpButton *microphoneSelection;
@property (weak) IBOutlet NSPopUpButton *speakerSelection;
@property (weak) IBOutlet NSPopUpButton *cameraSelection;

@property (strong, nonatomic) NSArray* connectedRecordingDevices;
@property (strong, nonatomic) NSArray* connectedPlaybackDevices;
@property (strong, nonatomic) NSArray* connectedVideoCaptureDevices;

@end

@implementation DeviceSelectionViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.preferredContentSize = NSMakeSize(500, 250);
    [self loadDevicesInPopUpButtons];
}

- (void)loadDevicesInPopUpButtons {
    [self.microphoneSelection removeAllItems];
    [self.speakerSelection removeAllItems];
    [self.cameraSelection removeAllItems];

    self.connectedRecordingDevices = [self.metaKit enumerateDevices:0];
    for (id device in self.connectedRecordingDevices) {
        [self.microphoneSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedPlaybackDevices = [self.metaKit enumerateDevices:1];
    for (id device in self.connectedPlaybackDevices) {
        [self.speakerSelection addItemWithTitle:([device deviceName])];
    }
    
    self.connectedVideoCaptureDevices = [self.metaKit enumerateDevices:3];
    for (id device in self.connectedVideoCaptureDevices) {
        [self.cameraSelection addItemWithTitle:([device deviceName])];
    // Populate the NSPopUpButtons with the enumerated device list
    }
}

- (void)rtcEngine:(MetaRtcEngineKit *)engine device:(NSString*) deviceId type:(MetaMediaDeviceType) deviceType stateChanged:(NSInteger) state {
    [self loadDevicesInPopUpButtons];
    // Repopulate NSPopUpButtons if a device is plugged in or removed
}

- (IBAction)didClickConfirmButton:(NSButton *)button {
    [self.metaKit setDevice:0 deviceId:[self.connectedRecordingDevices[self.microphoneSelection.indexOfSelectedItem] deviceId]];
    [self.metaKit setDevice:1 deviceId:[self.connectedPlaybackDevices[self.speakerSelection.indexOfSelectedItem] deviceId]];
    [self.metaKit setDevice:3 deviceId:[self.connectedVideoCaptureDevices[self.cameraSelection.indexOfSelectedItem] deviceId]];
    // Set the devices
    [self dismissViewController:self];
}

@end

File: One-to-One-Video/Meta-macOS-Tutorial-Swift-1to1/Meta-Mac-Tutorial-Swift/DeviceSelectionViewController.swift
//
//  DeviceSelectionViewController.swift
//  Meta Mac Tutorial Swift
//
//  Created by 3 on 2020/12/18.
//  Portions Copyright (c) 2020 meta-rti. All rights reserved.
//

import Cocoa
import MetaRTCFramework

class DeviceSelectionViewController: NSViewController {
    @IBOutlet weak var microphoneSelection: NSPopUpButton!
    @IBOutlet weak var speakerSelection: NSPopUpButton!
    @IBOutlet weak var cameraSelection: NSPopUpButton!
    
    var connectedRecordingDevices: [MetaRtcDeviceInfo]!
    var connectedPlaybackDevices: [MetaRtcDeviceInfo]!
    var connectedVideoCaptureDevices: [MetaRtcDeviceInfo]!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.preferredContentSize = NSMakeSize(500, 250)
        loadDevicesInPopUpButtons()
    }
    
    func loadDevicesInPopUpButtons() {
        microphoneSelection.removeAllItems()
        speakerSelection.removeAllItems()
        cameraSelection.removeAllItems()
    
        connectedRecordingDevices = MetaKit.enumerateDevices(.audioRecording)
        for device in connectedRecordingDevices {
            if let title = device.deviceName {
                microphoneSelection.addItem(withTitle: title)
            }
        }
    
        connectedPlaybackDevices = MetaKit.enumerateDevices(.audioPlayout)
        for device in connectedPlaybackDevices {
            if let title = device.deviceName {
                speakerSelection.addItem(withTitle: title)
            }
        }
        
        connectedVideoCaptureDevices = MetaKit.enumerateDevices(.videoCapture)
        for device in connectedVideoCaptureDevices {
            if let title = device.deviceName {
                cameraSelection.addItem(withTitle: title)
            }
        }
        // Populate the NSPopUpButtons with the enumerated device list
    }
    
    @IBAction func didClickConfirmButton(_ sender: NSButton) {
        if let deviceId = connectedRecordingDevices[microphoneSelection.indexOfSelectedItem].deviceId {
            MetaKit.setDevice(.audioRecording, deviceId: deviceId)
        }
        if let deviceId = connectedPlaybackDevices[speakerSelection.indexOfSelectedItem].deviceId {
            MetaKit.setDevice(.audioPlayout, deviceId: deviceId)
        }
        if let deviceId = connectedVideoCaptureDevices[cameraSelection.indexOfSelectedItem].deviceId {
            MetaKit.setDevice(.videoCapture, deviceId: deviceId)
        }
        // Set the devices
        self.dismiss(self)
    }
}

extension DeviceSelectionViewController: MetaRtcEngineDelegate {
    @objc(rtcEngine:device:type:stateChanged:) func rtcEngine(_ engine: MetaRtcEngineKit, device deviceId: String, type deviceType: MetaMediaDeviceType, stateChanged state: Int) {
        loadDevicesInPopUpButtons()
        // Repopulate NSPopUpButtons if a device is plugged in or removed
    }
}
